feat: counterfeit cycles model

ielashi opened the pull request.

ielashi committed:
feat: counterfeit cycles model

oggy-dfin commented on tla/counterfeit-cycles/counterfeit_cycles.tla:
```diff
@@ -0,0 +1,293 @@
+  
+  \* Choose two subnets.
+  \E sender, receiver \in SUBNETS: sender /= receiver
+  /\ subnets[sender].balance > 0
```
Strictly speaking, this is redundant because of the choice of amount below, but doesn't hurt obviously

ielashi replied:
Removed.

oggy-dfin commented on tla/counterfeit-cycles/counterfeit_cycles.tla:
```diff
@@ -0,0 +1,293 @@
+  
+  \* Send a transfer to the receiver.
+  /\ subnetMsgs' = Append(subnetMsgs, [
+        type |-> "transfer",
```
One trick I sometimes do is to define constants for these things, like `TRANSFER == "transfer"`. That way the parser at least complains about an unknown identifier if I mistype things. 
Just to mention, there's also another analysis tool for TLC:
https://apalache.informal.systems/
that comes with a type checker for TLA (and since recently also some kind of typed frontend language, but I haven't tried it yet).

oggy-dfin replied:
A style nitpick: I normally indent the clauses under a quantifier (i.e., everything below `\E sender...` in this definition) so I can see where they are bound, but that's obviously personal preference.

ielashi replied:
Nice trick. I added the constants and also updated the indentation per your suggestion.

oggy-dfin commented on tla/counterfeit-cycles/counterfeit_cycles.tla:
```diff
@@ -0,0 +1,293 @@
+     ])
+     
+  \* Subtract amount from sender's balance.
+  /\ subnets' = [subnets EXCEPT ![sender].balance = @ - amount]
```
I actually didn't know that you could do `![x].f = @` :D TIL!

ielashi replied:
Perhaps it was the programmer in me, but I wanted to put the subnet state into a single variable, and with the `@` it was super verbose.

oggy-dfin commented on tla/counterfeit-cycles/counterfeit_cycles.tla:
```diff
@@ -0,0 +1,293 @@
+  
+  \E sender, receiver \in SUBNETS: sender /= receiver
+
+  \* To keep the state space bounded, a subnet can make a dishonest
```
It may make more sense to limit the total amount of dishonest actions taken, instead of limiting the subnets to one dishonest action each; right now you would miss any attacks that would require the same subnet to do two dishonest things. I.e., have a counter of `total_dishonest_actions`, check that it's under some `MAX_DISHONEST_ACTIONS` and bump the counter here.

ielashi replied:
Agreed. Done.

oggy-dfin commented on tla/counterfeit-cycles/counterfeit_cycles.tla:
```diff
@@ -0,0 +1,293 @@
+        type |-> "transfer",
+        from |-> sender,
+        to |-> receiver,
+        \* Choose an amount that's greater than the balance.
```
I know we discussed potentially improving this but I don't have a really good suggestion for improvement here :( Obviously you need to limit the state space so you need to pick an amount from some finite set. You could pick an amount from `1..MAX_COUNTERFEIT_CYCLES` or something like that, that might be a bit nicer.

ielashi replied:
I agree. Interestingly enough, when I did this, I discovered a behavior that breaks one of the invariants. I'll follow up on this :)

oggy-dfin replied:
OK, you got my attention :) could it be an off-by-1 error somewhere? I guess that would be easy to miss with a +10 increase.

ielashi replied:
The issue we're running into here is that the model assumes that whether or not a transfer is "honest" is binary (i.e. a transfer is either honest or dishonest). This isn't exactly true. For example, a subnet with a balance of 2 can dishonestly try transferring 3 cycles, but as it's doing so it receives 1 cycle, and therefore the transfer now becomes "honest" as there are enough cycles available.

When we add +10, because that's more than the total supply, the example above never occurs. I think I need to think a bit harder about what the properties and invariants that we want to maintain, and what it really means for something to be "honest" - maybe quantifying what's "honest" isn't even relevant and can be removed 🤷‍♂️

oggy-dfin replied:
OK, that's interesting, I haven't thought about that case. I'd say that the transfer from your example is dishonest simply by definition, because an honest subnet would not do such a transfer. But I see that it violates the property that dishonest subnets are eventually detected. As you say maybe we can live without such a property, or maybe we can change the protocol to still detect such subnets. 

But I can only think of quite complicated solutions, something like: the ledger signs current balances together with a ledger time, and then the subnets provide their latest known signature of their balance with outgoing transfers, which the receiving subnet forwards to the ledger.  Eventually the ledger could do a reconciliation of all the records. But it's really complicated (and the tECDSA signatures are quite possibly too slow for this, and certified variables don't work in replicated mode), it would require unbounded storage on the ledger (unless we can come up with some garbage collection scheme). It also wouldn't detect the case where dishonest subnets send cycles to each other in invalid ways, but I think it's safe to say we will never be able to detect this reliably.

ielashi replied:
Right, for simplicity, I think the actual detection of whether or not a subnet behaved honestly isn't the goal here. The goal is to detect and stop the proliferation of counterfeit cycles.

I have updated the code now to address this, where I added `MAX_COUNTERFEIT_CYCLES` constant and updated the invariants accordingly.

oggy-dfin reviewed (commented):
Ran out of time for today, will continue tomorrow at some point

oggy-dfin commented on tla/counterfeit-cycles/counterfeit_cycles.tla:
```diff
@@ -0,0 +1,293 @@
+
+SubnetReceiveApprove == 
+  /\ Len(subnetMsgs) > 0
+  /\ LET msg == Head(subnetMsgs) IN
```
Could you be omitting some valid behaviors here because you impose an ordering? In particular, you only ever process the head of the `subnetMsgs` queue, but there's no guaranteed order in which different subnets would process these responses, and there's also no delivery order between messages sent by the other subnets and the ledger responses (actually, the interface spec doesn't even guarantee the delivery order for responses even between the ledger and the same subnet, though our implementation generally does). The same is actually true for the cycles-bearing messages between subnets (the receive order between subnets is not defined).

Unless you can justify why a queue is enough (which would actually likely be beneficial from an analysis perspective, as it would greatly decrease the number of possible actions), I'd suggest switching to sets. There are certain standard techniques that seem related to this (the queue could potentially be a so-called "prophecy variable" of how you are going to perform actions), but it can be a bit finicky.

ielashi replied:
I agree, and originally I didn't want to use a queue for `subnetMsgs`. The problem when I used a set was starvation - TLA would identify a sequence where a certain message is added to `subnetMsgs` but is never processed. Any ideas on how to address this?

I thought of trying to model the IC more precisely by having a separate queue per sender/receiver, but was worried of the complexity of that approach.

oggy-dfin replied:
> The problem when I used a set was starvation - TLA would identify a sequence where a certain message is added to subnetMsgs but is never processed. Any ideas on how to address this?

I'd have to look at the concrete trace, but it sounds suspicious to me that you had a starvation problem (I guess TLC complained of deadlock?) with sets, but not with lists. In general a set should be more "lenient", in the sense that you allow more actions and are thus less likely to deadlock.

If you can recreate the problem by running TLC from the command line, you should get a `counterfeit_cycles_TTrace_<something>.tla` file somewhere (probably also if you use the toolbox, but I'm not really familiar with how that handles problems) that contains the counterexample and that I can have a look at.

> I thought of trying to model the IC more precisely by having a separate queue per sender/receiver, but was worried of the complexity of that approach.

Using a set should overapproximate the real behaviors (I guess not a surprise). This is sound for safety properties (all of your properties except for the ledger detecting a malicious subnet are safety). The only downside is that the analysis might perform a lot worse, but premature optimization and all that, so I'd start with a set.

There's already [a model of streams](https://github.com/dfinity-lab/formal-models/blob/b5a7e6390f7140ffa827a55410cb681c2e2b2e73/tla/canister-migration/Migration.tla#L271) per subnet pair. It's far too detailed for your purposes, but I thought I'd mention it if useful for inspiration (I'd still try sets first, though)

ielashi replied:
One of the problems with using a set for `subnetMsgs` is that transfers messages are not unique and can therefore be lost if the same transfer (e.g. of 1 cycle from "s1" to "s2") are done in a row.

I've made the following changes to address this:

1. I added an `id` field to every message. Similar transfers will have different IDs and therefore both remain represented in the set of `subnetMsgs`.
2. I added a `MAX_TRANSFERS` constant, which limits the number of transfers that can be done. Otherwise, the introduction of the `id` field would make the state space infinite AFAIU. With that introduction, I had to turn off the check for dead locks as TLC assumes we've hit a deadlock once `MAX_TRANSFERS` has been made.
3. I changed the `msgs` that the ledger receives from a queue to a set of queues, where there's a queue per subnet. This is more representative of what happens on the IC. This isn't strictly required but I thought it's a more accurate representation.

oggy-dfin replied:
You could go with a multiset instead of a set, then you don't need IDs. If you do go with the ID solution, one thing you can do is an an idle transition roughly like this:
```
Idle == 
  /\ next_id >= MAX_TRANSFERS
  /\ vars' = vars
```

ielashi replied:
I went ahead with the idle transition and re-enabled the deadlock check. Thanks for the tip!

oggy-dfin commented on tla/counterfeit-cycles/counterfeit_cycles.tla:
```diff
@@ -0,0 +1,293 @@
+  
+  /\ UNCHANGED<<ledger>>
+
+SubnetDishonestSendTransfer ==
```
BTW what you've done here is a nifty and rather general trick; one general way of modeling corrupt nodes in distributed systems is to not corrupt their state, but just corrupt the messages they send. That way you can still state invariants over the state, as you've done. Just thought I'd mention it because I found it pretty cool you did that :)

ielashi replied:
Funny you mention that, because originally I did corrupt the state rather than sending a corrupt message, but later (exactly to your point) I realized that it was not possible to state invariants in that case. I'm glad I've stumbled into the right convention, thanks for confirming :)

oggy-dfin commented on tla/counterfeit-cycles/counterfeit_cycles.tla:
```diff
@@ -0,0 +1,293 @@
+   /\ Len(ledger.msgs) > 0
+   
+   /\ IF msg.amount > ledger.balances[msg.from] THEN
+        
+        
```
```suggestion
```

oggy-dfin commented on tla/counterfeit-cycles/counterfeit_cycles.tla:
```diff
@@ -0,0 +1,293 @@
+LedgerIdentifiesAllDishonestSubnets ==
+  {s \in SUBNETS: ~subnets[s].honest} = ledger.dishonestSubnets
+
+LedgerIdentifiesAllDishonestSubnetsEventually ==
```
Nit: this property allows the ledger to temporarily falsely suspect a subnet, which may not be desirable (it just says that the ledger _eventually_ doesn't "overshoot" with it's assessment of who's dishonest). If you still want to keep this property, you may want to complement it with an invariant that the set of suspects is a subset of the actually corrupt set.

ielashi replied:
Property removed.

oggy-dfin commented on tla/counterfeit-cycles/counterfeit_cycles.tla:
```diff
@@ -0,0 +1,293 @@
+  IF ks = {} THEN << >>
+  ELSE << f[CHOOSE k \in ks: TRUE] >> \o RecordToSeq(f, ks \ {CHOOSE k \in ks: TRUE})
+
+RECURSIVE SumSeq(_)
```
You could replace these next few definitions with a definition from a sorta-stdlib for TLA:
https://github.com/tlaplus/CommunityModules/blob/3fb30f52426b45cda44541b906ad9ebe16ce07c4/modules/FiniteSetsExt.tla#L40C1-L40C14
See also:
https://github.com/dfinity/formal-models/blob/master/tla/init-vscode.sh
to make this work out of the box (hopefully!) in VSCode.

ielashi replied:
I want to do that, but the method you're referring to is for a set, whereas here I'm dealing with records and sequences. For some reason I didn't see similar definitions on sequences, and records don't seem to have any helper methods?

oggy-dfin replied:
I think you can just do `MapThenSumSet(LAMBDA s: ledger.balances[s], SUBNETS)`?

ielashi replied:
I'm running into a strange issue. TLC is definitely pulling the community modules JAR, as I see this in the output:

```
arsing file /Users/ielashi/dev/dfinity/formal-models/tla/counterfeit-cycles/counterfeit_cycles.tla
Parsing file /private/var/folders/hr/p6hys22n5118fzkkwz5f49w00000gn/T/tlc-4870354577222093126/TLC.tla (jar:file:/Users/ielashi/dev/dfinity/formal-models/tla/tlatools/tla2tools.jar!/tla2sany/StandardModules/TLC.tla)
Parsing file /private/var/folders/hr/p6hys22n5118fzkkwz5f49w00000gn/T/tlc-4870354577222093126/Integers.tla (jar:file:/Users/ielashi/dev/dfinity/formal-models/tla/tlatools/tla2tools.jar!/tla2sany/StandardModules/Integers.tla)
Parsing file /private/var/folders/hr/p6hys22n5118fzkkwz5f49w00000gn/T/tlc-4870354577222093126/Sequences.tla (jar:file:/Users/ielashi/dev/dfinity/formal-models/tla/tlatools/tla2tools.jar!/tla2sany/StandardModules/Sequences.tla)
Parsing file /private/var/folders/hr/p6hys22n5118fzkkwz5f49w00000gn/T/tlc-4870354577222093126/SequencesExt.tla (jar:file:/Users/ielashi/dev/dfinity/formal-models/tla/tlatools/CommunityModules-deps.jar!/SequencesExt.tla)
Parsing file /private/var/folders/hr/p6hys22n5118fzkkwz5f49w00000gn/T/tlc-4870354577222093126/FiniteSets.tla (jar:file:/Users/ielashi/dev/dfinity/formal-models/tla/tlatools/tla2tools.jar!/tla2sany/StandardModules/FiniteSets.tla)
Parsing file /private/var/folders/hr/p6hys22n5118fzkkwz5f49w00000gn/T/tlc-4870354577222093126/FiniteSetsExt.tla (jar:file:/Users/ielashi/dev/dfinity/formal-models/tla/tlatools/CommunityModules-deps.jar!/FiniteSetsExt.tla)
Parsing file /private/var/folders/hr/p6hys22n5118fzkkwz5f49w00000gn/T/tlc-4870354577222093126/Naturals.tla (jar:file:/Users/ielashi/dev/dfinity/formal-models/tla/tlatools/tla2tools.jar!/tla2sany/StandardModules/Naturals.tla)
Parsing file /private/var/folders/hr/p6hys22n5118fzkkwz5f49w00000gn/T/tlc-4870354577222093126/Functions.tla (jar:file:/Users/ielashi/dev/dfinity/formal-models/tla/tlatools/CommunityModules-deps.jar!/Functions.tla)
Parsing file /private/var/folders/hr/p6hys22n5118fzkkwz5f49w00000gn/T/tlc-4870354577222093126/Folds.tla (jar:file:/Users/ielashi/dev/dfinity/formal-models/tla/tlatools/CommunityModules-deps.jar!/Folds.tla)
Semantic processing of module Naturals
Semantic processing of module Sequences
Semantic processing of module FiniteSets
Semantic processing of module TLC
Semantic processing of module Integers
Semantic processing of module Folds
Semantic processing of module Functions
Semantic processing of module FiniteSetsExt
```

But for some reason it can't find `MapThenSumSet`, even though it's in theory declared in module FiniteSetsExt that it has just processed. Weird...

oggy-dfin replied:
Ah git blame tells me this has only been added 4 months ago:
https://github.com/tlaplus/CommunityModules/blame/master/modules/FiniteSetsExt.tla
so the JAR is outdated. I've bumped up the version in master now so it should work if you merge it.

oggy-dfin commented on tla/counterfeit-cycles/counterfeit_cycles.tla:
```diff
@@ -0,0 +1,293 @@
+
+NoFakeCyclesAreCreated ==
+  \* The total supply of finalized cycles in the ledger remains constant.
+  /\ SumSeq(RecordToSeq(ledger.balances, SUBNETS)) = STARTING_BALANCE_PER_SUBNET * Cardinality(SUBNETS)
```
Nit: I'd define a `STARTING_SUPPLY ==  STARTING_BALANCE_PER_SUBNET * Cardinality(SUBNETS)`  somewhere

oggy-dfin replied:
Alternatively, you could define an "action invariant" such as this:
`SumSeq(RecordToSeq(ledger.balances, SUBNETS)) < SumSeq(RecordToSeq(ledger'.balances, SUBNETS))`
Like invariants, action invariants can still be processed fairly efficiently by TLC, unlike general safety and temporal properties which can slow the analysis down significantly (basically it's a local check on each edge of the state graph, rather than a graph search).

ielashi replied:
Done.

oggy-dfin reviewed (commented):
Completed a pass now. The two "real" issues that we should clarify are around modeling maliciousness and using queues vs sets (see the detailed comments). The rest is mostly just cosmetics comments, hopefully still useful!

ielashi committed:
pr comments

ielashi committed:
Remove empty lines

ielashi committed:
fix comment

ielashi committed:
empty lines

ielashi committed:
format

ielashi committed:
Replace ledger queue with a set of queues

ielashi committed:
Add cap for number of transfers

Adding a camp for the number of transfers helps keep the state space
smaller, and allows us to specify IDs for each transfer. Having IDs for
each transfer is really important as it serves two purposes:

1. Allows us to use sets in subnetMsgs, as transfers are now unique.
2. Helps with implementing the refund mechanism (not yet implemented)

ielashi committed:
add clarifying comment

ielashi committed:
Add ids to transfers

ielashi committed:
Use a set rather than a queue for subnetMsgs

ielashi committed:
Remove the concept of dishonestSubnets

ielashi committed:
trailing whitespace

ielashi committed:
Remove deadlock by adding idle transition

ielashi committed:
add back type invariants

ielashi committed:
Update balance in dishonest transfers

ielashi committed:
vary the amount of counterfeit cycles.

ielashi commented:
@oggy-dfin I believe at this point I've addressed the main issues you raised. The only remaining part to model are refunds. I'm not 100% sure if this feature is strictly necessary, and it will complicate the model, but I'll try modeling it and see how far I can go.

oggy-dfin commented on tla/counterfeit-cycles/counterfeit_cycles.tla:
```diff
@@ -0,0 +1,295 @@
+            \* Note that the balance can be negative if the subnet maliciously
+            \* sends more cycles than it has.
+
+            \* The balance of (finalized) cycles on the subnet.
```
The comment seems partly duplicated here.

oggy-dfin commented on tla/counterfeit-cycles/counterfeit_cycles.tla:
```diff
@@ -0,0 +1,295 @@
+
+SubnetsTypeOK ==
+  /\ \A s \in subnetMsgs: s \in SubnetMessages
+  /\ subnets \in [SUBNETS -> [
```
I think the way that TLC will evaluate this is to actually construct the set on the right. Not that I understand the first thing about their codebase, but it looks like it:
https://github.com/tlaplus/tlaplus/blob/fa4968f60a0a460da4b7e182330680ce93bd96aa/tlatools/org.lamport.tlatools/src/tlc2/tool/impl/Tool.java#L2591
(there are other mentions of OPCode_In, but roughly they seem to all be evaluating the right-hand side).
So if the analysis turns out to be slow, you might want to turn off the TypeOK checks for larger models, or rewriting the checks to be more verbose (like `DOMAIN subnets = SUBNETS /\ ...`).

oggy-dfin commented on tla/counterfeit-cycles/counterfeit_cycles.tla:
```diff
@@ -0,0 +1,295 @@
+  \E s \in SUBNETS: Len(subnets[s].msgsToLedger) > 0
+    /\ LET msg == Head(subnets[s].msgsToLedger) IN
+      \* Send message to ledger.
+      /\ ledger' = [ledger EXCEPT !["msgs"][s] = Append(@,
```
Actually, why do you need this transition? Couldn't the ledger read directly from `msgsToLedger`? I don't think you'd miss any relevant behaviors that way, but you'd have fewer transitions. I mean it doesn't hurt to have it from a soundness perspective, but it just seems to increase the state space needlessly.

oggy-dfin replied:
Also, nitpicking, but you could also use `!.msgs[s]` instead of `!["msgs"][s]` I think?

ielashi replied:
I agree that it's not strictly required, but as I started working on getting rid of it, I felt that the drawbacks of having a slightly higher state space don't outweigh the gain of modeling the process more accurately in this case, so I'm thinking of leaving it as-is. If you feel strongly otherwise, let me know and I'd be happy to remove that transition.

oggy-dfin replied:
I don't feel strongly about it, but if you need to speed the model checking up further at some point, getting rid of `ledger.msgs` would be my suggestion.

oggy-dfin commented on tla/counterfeit-cycles/counterfeit_cycles.tla:
```diff
@@ -0,0 +1,295 @@
+    /\ (\A s \in SUBNETS: Len(subnets[s].msgsToLedger) = 0)
+    /\ Cardinality(subnetMsgs) = 0
+  => \A sn \in SUBNETS: ledger.balances[sn] = subnets[sn].balance
+
```
I'd add a few checks such as
```suggestion
SanityCanGetApprovals == \A msg \in subnetMsgs: msg.type # APPROVE
SanityNoMoreThanOneMessage = Cardinality(subnetMsgs) < 2
```
And run the model to make sure that they are violated.

ielashi replied:
Verified that both properties when added are violated. I wonder if there are some type of TLA+ unit tests, where one could assert that the addition of a property leads to a violation. If I had more time on my hands I'd probably put something together that does this.

oggy-dfin replied:
I don't think that there are such tests. I'm personally a bit scared of the TLC code base :) but https://groups.google.com/g/tlaplus might be a place to discuss

oggy-dfin reviewed (approved):
LGTM overall now. There are a few comments where I think you can make things simpler, but overall seems sound to me (given the constraints of maintaining a finite state space).

ielashi committed:
.

ielashi committed:
rename ledger to tracker
